File: week3/승진/SW5650_핀볼 게임.py
import sys
from collections import deque

input = sys.stdin.readline

T = int(input().rstrip())




q = []

moves = [[-1,0],[1,0],[0,-1],[0,1]]

change = [[],[5,3,0,5],[3,5,1,5],[2,5,5,1],[5,2,5,0],[5,5,5,5]]









for test_case in range(1, T + 1):
    max_cnt = 0
    N = int(input().rstrip())
    maps = [[5 for _ in range(N + 2)]]
    hole ={6:[],7:[],8:[],9:[],10:[]}
    for i in range(N):
        tmp = [5]
        tmp.extend(list(map(int, input().rstrip().split())))
        for j,t in enumerate(tmp):
            if 6 <= t <= 10:
                hole[t].append([i + 1, j])
        tmp.append(5)
        maps.append(tmp)
    
    maps.append([5 for _ in range(N + 2)])
    
    print(hole)
    for m in maps:
        print(m)
    
    for i in range(1, N + 1):
        for j in range(1, N + 1):
            if maps[i][j] == 0:
                q.append((i, j, 0,0))
                q.append((i, j, 1,0))
                q.append((i, j, 2,0))
                q.append((i, j, 3,0))
                while q:
                    x, y, z, cnt = q.pop()
                    while True:
                        x, y = x + moves[z][0], y + moves[z][1]
                        if x == i and y == j:
                            break
                        
                        if maps[x][y] == 5 or maps[x][y] == -1:
                            break
                        elif 1 <= maps[x][y] < 5:
                            
                            z = change[maps[x][y]][z]
                            if z == 5:
                                cnt *= 2
                                cnt += 1
                                break
                            cnt += 1
                        elif 6 <= maps[x][y] <= 10:
                            tmp = hole[maps[x][y]]
                            for arr in tmp:
                                if arr != [x,y]:
                                    x, y = arr
                                    break                
                    if maps[x][y] == 5:
                        cnt *= 2
                        cnt += 1
                    
                    max_cnt = max(max_cnt, cnt)
    print(f'#{test_case} {max_cnt}')

File: week3/승진/SW5650_핀볼 게임.java
import java.util.*;

public class Solution {
    static class Pair {
        int x, y, z, cnt;

        Pair(int x, int y, int z, int cnt) {
            this.x = x;
            this.y = y;
            this.z = z;
            this.cnt = cnt;
        }
    }

    static int[][] moves = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    static int[][] change = {
            {}, {5, 3, 0, 5}, {3, 5, 1, 5}, {2, 5, 5, 1}, {5, 2, 5, 0}, {5, 5, 5, 5}
    };

    static int solve(int[][] maps, int N) {
        int max_cnt = 0;
        Map<Integer, List<int[]>> hole = new HashMap<>();
        for (int i = 1; i <= N; i++) {
            for (int j = 1; j <= N; j++) {
                if (maps[i][j] >= 6 && maps[i][j] <= 10) {
                    hole.computeIfAbsent(maps[i][j], k -> new ArrayList<>()).add(new int[]{i, j});
                }
            }
        }

        for (int i = 1; i <= N; i++) {
            for (int j = 1; j <= N; j++) {
                if (maps[i][j] == 0) {
                    for (int k = 0; k < 4; k++) {
                        int cnt = bfs(maps, N, i, j, k, hole);
                        max_cnt = Math.max(max_cnt, cnt);
                    }
                }
            }
        }
        return max_cnt;
    }

    static int bfs(int[][] maps, int N, int startX, int startY, int startZ, Map<Integer, List<int[]>> hole) {
        Deque<Pair> q = new ArrayDeque<>();
        q.push(new Pair(startX, startY, startZ, 0));

        while (!q.isEmpty()) {
            Pair pair = q.pop();
            int x = pair.x;
            int y = pair.y;
            int z = pair.z;
            int cnt = pair.cnt;

            while (true) {
                x += moves[z][0];
                y += moves[z][1];

                if (x == startX && y == startY) break;
                if (maps[x][y] == 5 || maps[x][y] == -1) break;
                if (maps[x][y] >= 1 && maps[x][y] < 5) {
                    z = change[maps[x][y]][z];
                    if (z == 5) {
                        cnt *= 2;
                        cnt++;
                        break;
                    }
                    cnt++;
                } else if (maps[x][y] >= 6 && maps[x][y] <= 10) {
                    List<int[]> tmp = hole.get(maps[x][y]);
                    for (int[] arr : tmp) {
                        if (arr[0] != x || arr[1] != y) {
                            x = arr[0];
                            y = arr[1];
                            break;
                        }
                    }
                }
            }

            if (maps[x][y] == 5) {
                cnt *= 2;
                cnt++;
            }

            return cnt;
        }

        return 0;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int T = Integer.parseInt(scanner.nextLine().trim());

        for (int test_case = 1; test_case <= T; test_case++) {
            int N = Integer.parseInt(scanner.nextLine().trim());
            int[][] maps = new int[N + 2][N + 2];
            for (int i = 0; i <= N + 1; i++) {
                Arrays.fill(maps[i], 5);
            }

            for (int i = 1; i <= N; i++) {
                String[] input = scanner.nextLine().split(" ");
                for (int j = 1; j <= N; j++) {
                    maps[i][j] = Integer.parseInt(input[j - 1]);
                }
            }

            int result = solve(maps, N);
            System.out.println("#" + test_case + " " + result);
        }
    }
}
